langchain4j/src/main/java/dev/langchain4j/service/ServiceOutputParser.java

```diff
@@ -22,67 +22,81 @@ public ServiceOutputParser(OutputParserFactory outputParserFactory) {
+        if (rawReturnClass == AiMessage.class) {
             return aiMessage;
         }
 
         String text = aiMessage.text();
-        if (returnType == String.class) {
+        if (rawReturnClass == String.class) {
             return text;
         }
 
-        Optional<OutputParser<?>> optionalOutputParser = outputParserFactory.get(returnType);
+        Optional<OutputParser<?>> optionalOutputParser = outputParserFactory.get(rawReturnClass, typeArgumentClass);
         if (optionalOutputParser.isPresent()) {
             return optionalOutputParser.get().parse(text);
         }
 
-        if (returnType == List.class) {
-            return asList(text.split("\n"));
-        }
+        return Json.fromJson(text, rawReturnClass);
+    }
 
-        if (returnType == Set.class) {
-            return new HashSet<>(asList(text.split("\n")));
-        }
+    public String outputFormatInstructions(Type returnType) {
 
-        return Json.fromJson(text, returnType);
-    }
+        if (typeHasRawClass(returnType, Result.class)) {
+            returnType = resolveFirstGenericParameterClass(returnType);
+        }
 
-    public String outputFormatInstructions(Class<?> returnType) {
+        // Explanation (which will make this a lot easier to understand):
+        // In the case of List<String> these two would be set like:
+        // rawClass: List.class
+        // typeArgumentClass: String.class
+        Class<?> rawClass = getRawClass(returnType);
+        Class<?> typeArgumentClass = TypeUtils.resolveFirstGenericParameterClass(returnType);
 
-        if (returnType == String.class
-                || returnType == AiMessage.class
-                || returnType == TokenStream.class
-                || returnType == Response.class) {
+        if (rawClass == String.class
+                || rawClass == AiMessage.class
+                || rawClass == TokenStream.class
+                || rawClass == Response.class) {
             return "";
         }
 
         if (returnType == void.class) {
             throw illegalConfiguration("Return type of method '%s' cannot be void");
         }
 
-        if (returnType.isEnum()) {
-            String formatInstructions = new EnumOutputParser(returnType.asSubclass(Enum.class)).formatInstructions();
-            return "\nYou must answer strictly with " + formatInstructions;
-        }
-
-        Optional<OutputParser<?>> outputParser = outputParserFactory.get(returnType);
+        Optional<OutputParser<?>> outputParser = outputParserFactory.get(rawClass, typeArgumentClass);
         if (outputParser.isPresent()) {
             String formatInstructions = outputParser.get().formatInstructions();
-            return "\nYou must answer strictly in the following format: " + formatInstructions;
-        }
 
-        if (returnType == List.class || returnType == Set.class) {
-            return "\nYou must put every item on a separate line.";
+            if (rawClass == List.class ||
```

Comment:
Perharps we should duplicate `\nYou must answer strictly in the following format: `message in all output parsers to make it more uniform and avoid this brittle `if` condition, WDYT?

Reply:
Actually I was already thinknig about that.

That indeed makes sense as it's concrete output parser's responsibility to construct entire instruction (for itself) on it's own (especially as there can differ from parser to parser). That would consequently further simplify `ServiceOutputParser`.

It's just I wasn't sure this is something to be done as part of this PR as scope of changes was quite large already...

If you think it's fine to do that as part of this PR then I can look further into it.

Reply:
You're right, let's not increase the scope of this PR, we can address it in a separate one